BCNM Royal Jelly Implementation

dallano opened the pull request:
**_I affirm:_**
- [x] I have paid attention to this example and will edit again if need be to not break the formatting, or I will be ignored
- [x] I have read and understood the [Contributing Guide](https://github.com/LandSandBoat/server/blob/base/CONTRIBUTING.md)
- [x] I've _**tested my code and the things my code has changed**_ since the last commit in the PR, and will test after any later commits

## What does this pull request do?

Implements BCNM Royal Jelly

Changes:
 - Removed Princess Jellies skill set.
 - Gave Princess Jellies weakness depending on their element
 - Added SCRIPT variable to pathfinding to allow the jellies to path to the center of the room (Can be used for a multitude of other mobs)
 - Slightly increased pre-existing spell cool down on Princes Jellies as they were casting faster than evidence found in youtube videos.
 - Removed some unused functions from code ported.
 - Increased the distance at which the jellies will cast on players to 20 yalms. (Previously set at 15 which would allow players to cast on jellies without any retaliation)
 - Slightly decreased the odds a Princess Jelly would cast bind by 10% to 20% (Review of past videos proved 33% was too high of odds - which was the previous rate on code from Setzor)
 - Implemented the drop tables for Royal Jelly which was taken from ffxiclopedia with a sample rate of ~400+
 - Cleaned up table formatting in Princess_Jelly.lua
 - Fixed Queen Jelly from spawning in the initiation of the battlefield
 - Moved some modifiers being added onto princess and queen luas into mob pools SQL
 - Added spell lists for both Princess and Queen Jelly

## Steps to test these changes
Enter BCNM Royal Jelly in Waughroon Shrine to test fight

dallano committed:
BCNM Royal Jelly Implementation

Co-Authored-By: Setz0r <setzor@gmail.com>

Basis and evidence for the following changes received from the following:
https://www.youtube.com/watch?v=pqMJou8WtYo&ab_channel=Pepriona
https://www.youtube.com/watch?v=UUykCuzhphQ&ab_channel=ffxikaiia
https://www.youtube.com/watch?v=R_gVC7uDLso&ab_channel=JohnCorona
https://www.youtube.com/watch?v=8xm-xvPjQkw&ab_channel=HeavyMetalWyrm
https://ffxiclopedia.fandom.com/wiki/Royal_Jelly_(BCNM)

A majority of the code handling Princess Jelly was ported from Setzor's implementation.

Changes:
Removed Princess Jellies skill set.
Gave Princess Jellies weakness depending on their element
Added SCRIPT variable to pathfinding to allow the jellies to path to the center of the room (Can be used for a multitude of other mobs)
Slightly increased pre-existing spell cool down on Princes Jellies as they were casting faster than evidence found in youtube videos.
Removed some unused functions from code ported.
Increased the distance at which the jellies will cast on players to 20 yalms. (Previously set at 15 which would allow players to cast on jellies without any retaliation)
Slightly decreased the odds a Princess Jelly would cast bind by 10% to 20% (Review of past videos proved 33% was too high of odds - which was the previous rate on code from Setzor)
Implemented the drop tables for Royal Jelly which was taken from ffxiclopedia with a sample rate of ~400+
Cleaned up table formatting in Princess_Jelly.lua
Fixed Queen Jelly from spawning in the initiation of the battlefield
Moved some modifiers being added onto princess and queen luas into mob pools SQL
Added spell lists for both Princess and Queen Jelly

dallano committed:
Style Fixing

Converted function(queen) to function(queenArg)

claywar commented on scripts/zones/Waughroon_Shrine/mobs/Princess_Jelly.lua:
```diff
@@ -0,0 +1,185 @@
+    {17367194, 17367195, 17367196, 17367197, 17367198, 17367199, 17367200, 17367201},
+}
+local elementalSpells =
+{
```
Pretty sure we have enums for these, and should be used in that case

claywar commented on scripts/zones/Waughroon_Shrine/mobs/Princess_Jelly.lua:
```diff
@@ -0,0 +1,185 @@
+require("scripts/globals/utils")
+-----------------------------------
+local princesses =
+{
```
While not necessary, I think for readability's sake, it might be beneficial to explicitly key these values for the BCNM ID

claywar commented on scripts/zones/Waughroon_Shrine/mobs/Princess_Jelly.lua:
```diff
@@ -0,0 +1,185 @@
+local function getDistanceFromCenter(bfNum,mob)
+    local pos = mob:getPos()
+
+    local centerX = centers[bfNum][1]
```
Would suggest just pulling the array as a single value, or assigning this directly in difX/Y/Z.  Seems a bit excessive

claywar commented on scripts/zones/Waughroon_Shrine/npcs/Armoury_Crate.lua:
```diff
@@ -116,6 +116,65 @@ local loot =
         },
     },
 
+     -- BCNM Royal Jelly
```
Assuming the rest of this script doesn't have a "pretty" formatting pattern, but would be nice to clean this up in the future.

dallano committed:
Royal Jelly Fixes

Created IDs for mobs to clean up file
Used enum values for spells
Fixed jellies not casting spells

dallano committed:
Update Princess_Jelly.lua

removed comment

TeoTwawki commented on scripts/zones/Waughroon_Shrine/mobs/Princess_Jelly.lua:
```diff
@@ -0,0 +1,172 @@
+    mob:setLocalVar('mobElement', math.random(1,8))
+    local element = mob:getLocalVar('mobElement')
+
+    if element == 1 then
```
Could use a keyed table instead of all this `elseif` conditionaling.

```lua
local mevaList =
{
    [0] = xi.mod.LIGHT_MEVA,
    [1] = xi.mod.WATER_MEVA,
    [2] = xi.mod.THUNDER_MEVA,
    [3] = xi.mod.EARTH_MEVA,
    [4] = xi.mod.WIND_MEVA,
    [5] = xi.mod.ICE_MEVA,
    [6] = xi.mod.FIRE_MEVA,
    [7] = xi.mod.DARK_MEVA
}
mob:addMod(mevaList[mob:getLocalVar('mobElement')], -250)
```
A variable not set with always return zero, so I just put light at the top as zero instead of an `else`.

edit: realized I don't even need the `element` variable, can get the localVar right there on the spot.

TeoTwawki commented on scripts/zones/Waughroon_Shrine/IDs.lua:
```diff
@@ -86,6 +86,50 @@ zones[xi.zone.WAUGHROON_SHRINE] =
+    {
+        [1] =
+        {
+            17367174,
```
since they are all sequential in the 3 sets, could we possibly use just the 1st of each set and offset from there?

TeoTwawki replied:
actually, could just offset from the queens IDs

dallano committed:
Royal Jelly Updates

- added checks for princess jellies based off queen's
- added a meva list used for calculating magic eva down

TeoTwawki commented:
needs rebased
